domain/secretbackend/service/service.go

```diff
@@ -0,0 +1,580 @@
+
+	cfgToApply := make(map[string]interface{})
+	for k, v := range existing.Config {
+		cfgToApply[k] = v
+	}
+	for k, v := range params.Config {
+		cfgToApply[k] = v
+	}
+	for _, k := range params.Reset {
+		delete(cfgToApply, k)
+	}
+	configValidator, ok := p.(provider.ProviderConfig)
+	if ok {
+		defaults := configValidator.ConfigDefaults()
+		for _, k := range params.Reset {
+			if defaultVal, ok := defaults[k]; ok {
+				cfgToApply[k] = defaultVal
+			}
+		}
+		err = configValidator.ValidateConfig(convertConfigToAny(existing.Config), cfgToApply)
+		if err != nil {
+			return fmt.Errorf("%w: config for provider %q: %w", secretbackenderrors.NotValid, existing.BackendType, err)
+		}
+	}
+	if !params.Force {
+		if err := pingBackend(p, cfgToApply); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	params.Config = convertConfigToString(cfgToApply)
+
+	if params.TokenRotateInterval != nil && *params.TokenRotateInterval > 0 {
+		if !provider.HasAuthRefresh(p) {
+			return errors.NotSupportedf("token refresh on secret backend of type %q", p.Type())
+		}
+		params.NextRotateTime, err = coresecrets.NextBackendRotateTime(s.clock.Now(), *params.TokenRotateInterval)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	_, err = s.st.UpdateSecretBackend(ctx, params.UpdateSecretBackendParams)
+	return errors.Trace(err)
+}
+
+// DeleteSecretBackend deletes a secret backend.
+func (s *Service) DeleteSecretBackend(ctx context.Context, backendID string, force bool) error {
+	return s.st.DeleteSecretBackend(ctx, backendID, force)
+}
+
+// GetSecretBackendByName returns the secret backend for the given backend name.
+func (s *Service) GetSecretBackendByName(ctx context.Context, name string) (*coresecrets.SecretBackend, error) {
+	sb, err := s.st.GetSecretBackend(ctx, secretbackend.BackendIdentifier{Name: name})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &coresecrets.SecretBackend{
+		ID:                  sb.ID,
+		Name:                sb.Name,
+		BackendType:         sb.BackendType,
+		TokenRotateInterval: sb.TokenRotateInterval,
+		Config:              convertConfigToAny(sb.Config),
+	}, nil
+}
+
+// RotateBackendToken rotates the token for the given secret backend.
+func (s *Service) RotateBackendToken(ctx context.Context, backendID string) error {
+	backendInfo, err := s.st.GetSecretBackend(ctx,
+		secretbackend.BackendIdentifier{ID: backendID},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	p, err := s.registry(backendInfo.BackendType)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !provider.HasAuthRefresh(p) {
+		return nil
+	}
+
+	if backendInfo.TokenRotateInterval == nil || *backendInfo.TokenRotateInterval == 0 {
+		s.logger.Debugf("not rotating token for secret backend %q", backendInfo.Name)
+		return nil
+	}
+
+	s.logger.Debugf("refresh token for backend %v", backendInfo.Name)
+	cfg := provider.BackendConfig{
+		BackendType: backendInfo.BackendType,
+		Config:      convertConfigToAny(backendInfo.Config),
+	}
```

Comment:
I think we do this a bit - we could have a helper to convert a domain backend to a provider backend

Reply:
Not quite sure if it's worth to do it.
The provider.BackendConfig only has two fields.
And provider.BackendConfig is just subset of the domain backendInfo, so backendInfo.ToBackendConfig() seems very weird.